src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,1311 @@
+    uint64_t count = 0;
+    mDatabase.getSession() << query, into(count), use(ledgers.first()),
+        use(ledgers.last());
+    return count;
+}
+
+void
+LedgerStateRoot::deleteObjectsModifiedOnOrAfterLedger(
+        uint32_t ledger) const
+{
+    return mImpl->deleteObjectsModifiedOnOrAfterLedger(ledger);
+}
+
+void
+LedgerStateRoot::Impl::deleteObjectsModifiedOnOrAfterLedger(
+        uint32_t ledger) const
+{
+    using namespace soci;
+
+    mCache->clear();
+
+    {
+        std::string query =
+            "DELETE FROM signers WHERE accountid IN"
+            " (SELECT accountid FROM accounts WHERE lastmodified >= :v1)";
+        mDatabase.getSession() << query, use(ledger);
+    }
+
+    for (auto let : {ACCOUNT, DATA, TRUSTLINE, OFFER})
+    {
+        std::string query =
+            "DELETE FROM " + tableFromLedgerEntryType(let) +
+            " WHERE lastmodified >= :v1";
+        mDatabase.getSession() << query, use(ledger);
+    }
+}
+
+void
+LedgerStateRoot::dropAccounts()
+{
+    mImpl->dropAccounts();
+}
+
+void
+LedgerStateRoot::dropData()
+{
+    mImpl->dropData();
+}
+
+void
+LedgerStateRoot::dropOffers()
+{
+    mImpl->dropOffers();
+}
+
+void
+LedgerStateRoot::dropTrustLines()
+{
+    mImpl->dropTrustLines();
+}
+
+std::map<LedgerKey, LedgerEntry>
+LedgerStateRoot::getAllOffers()
+{
+    return mImpl->getAllOffers();
+}
+
+std::map<LedgerKey, LedgerEntry>
+LedgerStateRoot::Impl::getAllOffers()
+{
+    auto offers = loadAllOffers();
+    std::map<LedgerKey, LedgerEntry> offersByAccount;
+    for (auto const& offer : offers)
+    {
+        offersByAccount[LedgerEntryKey(offer)] = offer;
+    }
+    return offersByAccount;
+}
+
+std::shared_ptr<LedgerEntry>
+LedgerStateRoot::getBestOffer(Asset const& buying, Asset const& selling,
+                              std::set<LedgerKey>&& exclude)
+{
+    return mImpl->getBestOffer(buying, selling, std::move(exclude));
+}
+
+std::shared_ptr<LedgerEntry>
+LedgerStateRoot::Impl::getBestOffer(Asset const& buying, Asset const& selling,
+                                    std::set<LedgerKey>&& exclude)
+{
+    std::string cacheKey = binToHex(xdr::xdr_to_opaque(buying))
+                         + binToHex(xdr::xdr_to_opaque(selling));
+    if (!mBestOffersCache->exists(cacheKey))
+    {
+        mBestOffersCache->put(cacheKey, {{}, false});
+    }
+    auto& cached = mBestOffersCache->get(cacheKey);
+
+    auto& offers = cached.first;
+    for (auto const& offer : offers)
+    {
+        auto key = LedgerEntryKey(offer);
+        if (exclude.find(key) == exclude.end())
+        {
+            return std::make_shared<LedgerEntry>(offer);
+        }
+    }
+
+    while (!cached.second)
+    {
+        auto newOffers = loadBestOffers(buying, selling, 5, offers.size());
+        if (newOffers.size() < 5)
+        {
+            cached.second = true;
+        }
+
+        offers.insert(offers.end(), newOffers.begin(), newOffers.end());
+        for (auto const& offer : newOffers)
```

Comment:
we may want to do a pass at those as well: we copy for no reason (could iterate over the last `N` items from offers)